src/core/Akka.Streams.Tests/Dsl/SinkForeachAsyncSpec.cs

```diff
@@ -0,0 +1,259 @@
+
+            latch[1].CountDown();
+            probe.ExpectMsg(1);
+            latch[2].CountDown();
+            probe.ExpectMsg(2);
+            latch[3].CountDown();
+            probe.ExpectMsg(3);
+            latch[4].CountDown();
+            probe.ExpectMsg(4);
+
+            (await p.ShouldCompleteWithin(TimeSpan.FromSeconds(4))).Should().Be(Done.Instance);
+        }
+
+        [Fact]
+        public async Task A_ForeachAsync_must_back_pressure_upstream_elements_when_downstream_is_slow()
+        {
+            var probe = CreateTestProbe();
+            var latch = Enumerable.Range(1, 4)
+                .Select(i => (i, new TestLatch(1)))
+                .ToDictionary(t => t.i, t => t.Item2);
+
+            var sink = Sink.ForEachAsync<Func<int>>(1, async n =>
+            {
+                latch[n()].Ready(RemainingOrDefault);
+                probe.Ref.Tell(n());
+                await Task.Delay(2000);
+            });
+
+            var oneCalled = false;
+            var twoCalled = false;
+            var threeCalled = false;
+            var fourCalled = false;
+
+            int One()
+            {
+                oneCalled = true;
+                return 1;
+            }
+
+            int Two()
+            {
+                twoCalled = true;
+                return 2;
+            }
+
+            int Three()
+            {
+                threeCalled = true;
+                return 3;
+            }
+
+            int Four()
+            {
+                fourCalled = true;
+                return 4;
+            }
+
+            var p = Source.From(new List<Func<int>> { One, Two, Three, Four }).RunWith(sink, Materializer);
+
+            latch[1].CountDown();
+            probe.ExpectMsg(1);
+
+            twoCalled.ShouldBeFalse();
+            threeCalled.ShouldBeFalse();
+            fourCalled.ShouldBeFalse();
+
+            probe.ExpectNoMsg(TimeSpan.FromSeconds(2));
+
+            latch[2].CountDown();
+            probe.ExpectMsg(2);
+
+            threeCalled.ShouldBeFalse();
+            fourCalled.ShouldBeFalse();
+
+            probe.ExpectNoMsg(TimeSpan.FromSeconds(2));
+
+            latch[3].CountDown();
+            probe.ExpectMsg(3);
+
+            fourCalled.ShouldBeFalse();
+
+            probe.ExpectNoMsg(TimeSpan.FromSeconds(2));
+
+            latch[4].CountDown();
+            probe.ExpectMsg(4);
+
+            (await p.ShouldCompleteWithin(TimeSpan.FromSeconds(4))).Should().Be(Done.Instance);
+
+            oneCalled.ShouldBeTrue();
+            twoCalled.ShouldBeTrue();
+            threeCalled.ShouldBeTrue();
+            fourCalled.ShouldBeTrue();
+        }
+
+        [Fact]
+        public async Task A_ForeachAsync_must_produce_elements_in_the_order_they_are_ready()
+        {
+            var probe = CreateTestProbe();
+            var latch = Enumerable.Range(1, 4)
+                .Select(i => (i, new AsyncCountdownEvent(1)))
```

Comment:
I had to used `AsyncCountdownEvent` directly in a bunch of tests because `TestLatch.Ready` currently blocks.